app/scripts/metamask-controller.js

```diff
@@ -2723,6 +2724,34 @@ export default class MetamaskController extends EventEmitter {
     return keyring.isUnlocked();
   }
 
+  /**
+   * Attempts to communicate with the hardware device and returns true if successful
+   *
+   * @param deviceName
+   * @param hdPath
+   * @returns {Promise<bool>}
+   */
+  async isDeviceAccessible(deviceName, hdPath) {
+    const keyring = await this.getKeyringForDevice(deviceName);
+    let status = false;
+    // ledger state becomes stale and must be explicitly accessed
+    if (keyring.type === HardwareKeyringTypes.ledger) {
+      try {
+        // unlock creates app and then attempts to read address from device
+        // attemptMakeApp will still return true when device locked
+        await keyring.unlock(hdPath, false);
```

Comment:
I think this is similar to what `this.checkHardwareStatus` does already, isn't it? Could this `.unlock` call still return false positives?

This is what the ledger does when `.unlock` is called.

```js
  isUnlocked () {
    return Boolean(this.hdk && this.hdk.publicKey)
  }
  
  unlock (hdPath, updateHdk = true) {
    if (this.isUnlocked() && !hdPath) {
      return Promise.resolve('already unlocked')
    }
```

Reply:
Maybe the one difference is that `checkHardwareStatus` doesn't have a try/catch. Have you experimented with adding a try/catch there to see if that fixes the issue?

Reply:
`unlock` only shortcuts if you do not provide the `hdPath`. When the `hdPath` is provided, it sends a `ledger-unlock` message to the hardware wallet:

```  unlock (hdPath, updateHdk = true) {
    if (this.isUnlocked() && !hdPath) {
      return Promise.resolve('already unlocked')
    }
    const path = hdPath ? this._toLedgerPath(hdPath) : this.hdPath
    return new Promise((resolve, reject) => {
      this._sendMessage({
        action: 'ledger-unlock',
        params: {
          hdPath: path,
        },
      },
      ({ success, payload }) => {
        if (success) {
          if (updateHdk) {
            this.hdk.publicKey = Buffer.from(payload.publicKey, 'hex')
            this.hdk.chainCode = Buffer.from(payload.chainCode, 'hex')
          }
          resolve(payload.address)
        } else {
          reject(payload.error || new Error('Unknown error'))
        }
      })
    })
  }```
  
  
  The `unlock` implementation, will then `makeApp` (doesn't re-establich connection) and then retrieves the address for the  `hdPath`, which re-establishes connection and throws an error if the device is locked:
  
  ```{
        key: 'unlock',
        value: async function unlock(replyAction, hdPath, messageId) {
            try {
                await this.makeApp();
                var res = await this.app.getAddress(hdPath, false, true);
                this.sendMessageToExtension({
                    action: replyAction,
                    success: true,
                    payload: res,
                    messageId: messageId
                });
            } catch (err) {
                var e = this.ledgerErrToMessage(err);
                this.sendMessageToExtension({
                    action: replyAction,
                    success: false,
                    payload: { error: e },
                    messageId: messageId
                });
            } finally {
                if (this.transportType !== 'ledgerLive') {
                    this.cleanUp();
                }
            }
        }
    }```